mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorRewriting.cpp

```diff
@@ -842,6 +844,31 @@ struct PrintRewriter : public OpRewritePattern<PrintOp> {
     rewriter.create<vector::PrintOp>(loc, vector::PrintPunctuation::Close);
   }
 
+  // Helper method to print the ending of memref (no punctuation after the last
+  // element).
+  static void printEnding(PatternRewriter &rewriter, Location loc, Value indVar,
+                          Value step, Value upperBound, bool isComplex,
+                          Value val) {
+    auto bound = rewriter.create<arith::AddIOp>(loc, indVar, step);
+    Value cond = rewriter.create<arith::CmpIOp>(loc, arith::CmpIPredicate::eq,
+                                                bound, upperBound);
+    scf::IfOp ifOp = rewriter.create<scf::IfOp>(loc, cond, /*else*/ true);
+    if (isComplex) {
```

Comment:
something does not feel right
why is the complex test inside the print end?

Reply:
Use this

     // Actual contents printing.
      auto val = rewriter.create<memref::LoadOp>(loc, vec, idxs);
      if (llvm::isa<ComplexType>(val.getType())) {
        // Since the vector dialect does not support complex types in any op,
        // we split those into (real, imag) pairs here.
        Value real = rewriter.create<complex::ReOp>(loc, val);
        Value imag = rewriter.create<complex::ImOp>(loc, val);
        rewriter.create<vector::PrintOp>(loc, vector::PrintPunctuation::Open);
        rewriter.create<vector::PrintOp>(loc, real,
                                         vector::PrintPunctuation::Comma);
        rewriter.create<vector::PrintOp>(loc, imag,
                                         vector::PrintPunctuation::Close);
      } else {
        rewriter.create<vector::PrintOp>(loc, val, vector::PrintPunctuation::NoPunctuation);
      }
      // YOUR IF I < SIZE -1 LOGIC HERE!
      rewriter.create<vector::PrintOp>(loc, vector::PrintPunctuation::Comma);

and then generate the IF at the marked place

Reply:
// Terminating comma (except at end).
      auto bound = rewriter.create<arith::AddIOp>(loc, idxs.back(), step);
      Value cond = rewriter.create<arith::CmpIOp>(loc, arith::CmpIPredicate::ne, bound, size);
      scf::IfOp ifOp = rewriter.create<scf::IfOp>(loc, cond, /*else*/ true);
      rewriter.setInsertionPointToStart(&ifOp.getThenRegion().front());
      rewriter.create<vector::PrintOp>(loc, vector::PrintPunctuation::Comma);